op-batcher/batcher/driver.go

```diff
@@ -465,68 +494,54 @@ func (l *BatchSubmitter) mainLoop(ctx context.Context, receiptsCh chan txmgr.TxR
-			queueCancel()
-			if err := queue.Wait(); err != nil {
-				if !errors.Is(err, context.Canceled) {
-					l.Log.Error("error waiting for transactions to complete", "err", err)
-				}
-			}
-			throttlingLoopCancel()
-			receiptsLoopCancel()
-			l.wg.Done()
-			l.Log.Warn("main loop returning")
+			l.Log.Warn("readloop returning")
 			return
 		}
 	}
 }
 
-// processReceiptsLoop handles transaction receipts from the DA layer
-func (l *BatchSubmitter) processReceiptsLoop(ctx context.Context, receiptsCh chan txmgr.TxReceipt[txRef]) {
-	defer l.wg.Done()
+// receiptsLoop handles transaction receipts from the DA layer
+func (l *BatchSubmitter) receiptsLoop(wg *sync.WaitGroup, receiptsCh chan txmgr.TxReceipt[txRef]) {
+	defer wg.Done()
 	l.Log.Info("Starting receipts processing loop")
-	for {
-		select {
-		case r := <-receiptsCh:
-			if errors.Is(r.Err, txpool.ErrAlreadyReserved) && l.txpoolState == TxpoolGood {
-				l.setTxPoolState(TxpoolBlocked, r.ID.isBlob)
-				l.Log.Warn("incompatible tx in txpool", "id", r.ID, "is_blob", r.ID.isBlob)
-			} else if r.ID.isCancel && l.txpoolState == TxpoolCancelPending {
-				// Set state to TxpoolGood even if the cancellation transaction ended in error
-				// since the stuck transaction could have cleared while we were waiting.
-				l.setTxPoolState(TxpoolGood, l.txpoolBlockedBlob)
-				l.Log.Info("txpool may no longer be blocked", "err", r.Err)
-			}
-			l.Log.Info("Handling receipt", "id", r.ID)
-			l.handleReceipt(r)
-		case <-ctx.Done():
-			l.Log.Info("Receipt processing loop done")
-			return
+	for r := range receiptsCh {
+
+		if errors.Is(r.Err, txpool.ErrAlreadyReserved) && l.txpoolState == TxpoolGood {
+			l.setTxPoolState(TxpoolBlocked, r.ID.isBlob)
+			l.Log.Warn("incompatible tx in txpool", "id", r.ID, "is_blob", r.ID.isBlob)
+		} else if r.ID.isCancel && l.txpoolState == TxpoolCancelPending {
+			// Set state to TxpoolGood even if the cancellation transaction ended in error
+			// since the stuck transaction could have cleared while we were waiting.
+			l.setTxPoolState(TxpoolGood, l.txpoolBlockedBlob)
+			l.Log.Info("txpool may no longer be blocked", "err", r.Err)
 		}
+		l.Log.Info("Handling receipt", "id", r.ID)
+		l.handleReceipt(r)
 	}
+	l.Log.Info("Receipt processing loop returning")
 }
 
 // throttlingLoop monitors the backlog in bytes we need to make available, and appropriately enables or disables
 // throttling of incoming data prevent the backlog from growing too large. By looping & calling the miner API setter
 // continuously, we ensure the engine currently in use is always going to be reset to the proper throttling settings
 // even in the event of sequencer failover.
-func (l *BatchSubmitter) throttlingLoop(ctx context.Context) {
-	defer l.wg.Done()
+func (l *BatchSubmitter) throttlingLoop(wg *sync.WaitGroup, pendingBytesUpdated chan int64) {
+	defer wg.Done()
 	l.Log.Info("Starting DA throttling loop")
-	ticker := time.NewTicker(l.Config.ThrottleInterval)
-	defer ticker.Stop()
```

Comment:
By removing the ticker, we only "retry" RPC calls to the sequencer when we get signalled by the `readLoop`, rather than on the next tick. And when we get that signal, we reevaluate the need to make the RPC calls and may not make them after all (so it's not really a retry). We could install some generic retry logic in here, to retry immediately (a limited number of times) if there is an RPC error.

Reply:
I went ahead with that here cc26ac9f418adc2f9e1bb98d534814b250752d96